File: scripts/youngcha/producers/crucible.js
const heatlib = require("libraries/heatlib");
const graphLib = require("libraries/graphlib");
const crucibleLib = require("libraries/cruciblelib");

let crublankobj = graphLib.init();
graphLib.addGraph(crublankobj, heatlib.baseTypesHeat.heatConnector);
graphLib.addGraph(crublankobj, crucibleLib.baseTypes.crucibleConnector);


const randomPos=[
	{x:0,y:0},
	{x:-1.6,y:1.6},
	{x:-1.6,y:-1.6},
	{x:1.6,y:-1.6},
	{x:-1.6,y:-1.6},
	{x:0,y:0},
		];


const crucible = graphLib.finaliseExtend(Block, Building, "crucible", crublankobj, {
    load() {
        this.super$load();
        this.liquidsprite = Core.atlas.find(this.name + "-liquid");
        this.bottom = Core.atlas.find(this.name + "-base");
		this.floorSprite = Core.atlas.find(this.name + "-floor");
        this.solidItem = Core.atlas.find(this.name + "-solid");
		this.solidItemStrip = Core.atlas.find(this.name + "-solidstrip");
		this.heatSprite = Core.atlas.find(this.name + "-heat");
    },
}, {
    updatePost() {},
    draw() {
        let temp = this.getGraphConnector("heat graph").getTemp();
        let dex = this.getGraphConnector("crucible graph").getTileIndex();
        let tileindex = crucibleLib.tileIndexMap[dex];
		Draw.rect(crucible.floorSprite, this.x, this.y, 8, 8, 0);
		this.drawContents(this.getGraphConnector("crucible graph"),crucible.liquidsprite,tileindex,crucible.solidItem,crucible.solidItemStrip);
		crucibleLib.drawTile(crucible.bottom, this.x, this.y, 8, 8, 0, tileindex);
        heatlib.drawHeat(crucibleLib.getRegion(crucible.heatSprite, tileindex), this.x, this.y, 0, temp);
        this.drawTeamTop();
    },

    //move out later
    //whether to accept the item? or the accepting of the item
    acceptItem(source, item) {
        let dex = this.getGraphConnector("crucible graph");
        return dex.canContainMore(1) && dex.getMelt(item);
    },

    //nvm thats handle item
    handleItem(source, item) {
        let dex = this.getGraphConnector("crucible graph");
        dex.addItem(item);
    },
	
	drawContents(crucgraph,liquidsprite,tindex, itemsprite, itemsprite2){
		let cc = crucgraph.getContained();
		if(!cc || cc.length==0){return;}
		
		let col = {r:0,g:0,b:0};
		let tliquid = 0;
		let fraction = crucgraph.liquidcap/crucgraph.getTotalLiquidCapacity();
		for(let i = 0;i<cc.length;i++){
			let itmcol = Color.valueOf("555555");
			if(cc[i].meltedRatio>0){
				let liquidvol = cc[i].meltedRatio*cc[i].volume;
				tliquid +=liquidvol;
				if(cc[i].item){
					itmcol = cc[i].item.color;
				}
				col.r +=itmcol.r*liquidvol;
				col.g +=itmcol.g*liquidvol;
				col.b +=itmcol.b*liquidvol;
			}
		}
		if(tliquid>0){
			let invt = 1.0/tliquid;
			Draw.color(new Color(col.r*invt,col.g*invt,col.b*invt), Mathf.clamp(tliquid*fraction*3.0));	
			Draw.rect(crucibleLib.getRegion(liquidsprite, tindex), this.x, this.y, 8, 8, 0);
		}
		for(let i = 0;i<cc.length;i++){
			if(cc[i].meltedRatio<1){
				let itmcol = Color.valueOf("555555");
				if(cc[i].item){
					itmcol = cc[i].item.color;
				}
				Draw.color(itmcol);	
				let ddd = (1-cc[i].meltedRatio)*cc[i].volume*fraction;
				if(ddd>1){
					Draw.rect(heatlib.getRegion(itemsprite2,Math.floor(ddd)-1,6,1), this.x, this.y, 0);
				}
				if(ddd>0.1){
					let siz = 8.0*(ddd%1.0);
					Draw.rect(itemsprite, randomPos[Math.min(Math.floor(ddd),5)].x + this.x, randomPos[Math.min(Math.floor(ddd),5)].y + this.y,siz,siz);
				}
			}
		}
		Draw.color();	
	}
	

});
crucible.update = true;
crucible.rotate = true;
crucible.solid = true;
crucible.getGraphConnectorBlock("crucible graph").setAccept([1, 1, 1, 1]);
crucible.getGraphConnectorBlock("heat graph").setAccept([1, 1, 1, 1]);
crucible.getGraphConnectorBlock("heat graph").getBaseHeatConductivity(0.2);
crucible.getGraphConnectorBlock("heat graph").getBaseHeatCapacity(75); //maybe make it dynamic later...
crucible.getGraphConnectorBlock("heat graph").setBaseHeatRadiativity(0.006);

File: main/src/unity/world/blocks/production/Crucible.java
package unity.world.blocks.production;

import arc.graphics.*;
import arc.graphics.g2d.*;
import arc.math.*;
import arc.scene.ui.layout.*;
import arc.struct.*;
import mindustry.gen.*;
import mindustry.type.*;
import mindustry.ui.*;
import unity.gen.*;
import unity.graphics.*;
import unity.ui.dialogs.*;
import unity.util.*;
import unity.world.blocks.*;
import unity.world.graph.*;
import unity.world.meta.*;
import unity.world.modules.*;

import static arc.Core.*;

public class Crucible extends GraphBlock{
    CrucibleGraph viewPos;
    private static final long[] randomPos = new long[]{
        SVec2.construct(0f, 0f),
        SVec2.construct(-1.6f, 1.6f),
        SVec2.construct(-1.6f, -1.6f),
        SVec2.construct(1.6f, -1.6f),
        SVec2.construct(-1.6f, -1.6f),
        SVec2.construct(0f, 0f)
    };

    public TextureRegion[] liquidRegions, baseRegions, roofRegions, solidItemStrips, heatRegions;
    public TextureRegion floorRegion, solidItem;

    public Crucible(String name){
        super(name);

        configurable = solid = true;
    }

    @Override
    public void load(){
        super.load();

        liquidRegions = GraphicUtils.getRegions(liquidRegion, 12, 4);
        baseRegions = GraphicUtils.getRegions(atlas.find(name + "-base"), 12, 4);
        floorRegion = atlas.find(name + "-floor");
        roofRegions = GraphicUtils.getRegions(atlas.find(name + "-roof"), 12, 4);

        solidItem = atlas.find(name + "-solid");
        solidItemStrips = GraphicUtils.getRegions(atlas.find(name + "-solidstrip"), 6, 1);
        heatRegions = GraphicUtils.getRegions(heatRegion, 12, 4);
    }

    public class CrucibleBuild extends GraphBuild{
        final Color color = Color.clear.cpy();

        @Override
        public void buildConfiguration(Table table){
            table.button(Tex.whiteui, Styles.clearTransi, 50f, new CrucibleDialog(this)::show).size(50f).get().getStyle().imageUp = Icon.chartBar;
            table.button(Tex.whiteui, Styles.clearTransi, 50f, () -> configure(0)).size(50f).get().getStyle().imageUp = Icon.eye;
        }

        @Override
        public void configured(Unit builder, Object value){
            CrucibleGraph thisG = crucible().getNetwork();
            viewPos = viewPos == thisG ? null : thisG;
        }

        @Override
        public void drawConfigure(){}

        @Override
        public void draw(){
            GraphCrucibleModule dex = crucible();
            byte tileIndex = UnityDrawf.tileMap[dex.tilingIndex];

            if(viewPos == dex.getNetwork()){
                Draw.rect(floorRegion, x, y, 8f, 8f);
                drawContents(dex, tileIndex);

                Draw.rect(baseRegions[tileIndex], x, y, 8f, 8f, 4f, 4f, 0f);
                UnityDrawf.drawHeat(heatRegions[tileIndex], x, y, 0f, heat().getTemp());
            }else{
                Draw.rect(roofRegions[tileIndex], x, y, 8f, 8f, 4f, 4f, 0f);
            }

            drawTeamTop();
        }

        @Override
        public boolean acceptItem(Building source, Item item){
            return crucible().canContainMore(1f) && MeltInfo.map.containsKey(item);
        }

        @Override
        public void handleItem(Building source, Item item){
            crucible().addItem(item);
        }

        protected void drawContents(GraphCrucibleModule crucGraph, int tIndex){
            color.set(0f, 0f, 0f);
            Seq<CrucibleData> cc = crucGraph.getContained();

            if(cc.isEmpty()) return;

            float tLiquid = 0f;
            float fraction = crucGraph.liquidCap / crucGraph.getTotalLiquidCapacity();

            for(var i : cc){
                if(i.meltedRatio > 0f){
                    float liquidVol = i.meltedRatio * i.volume;
                    tLiquid += liquidVol;
                    Color itemCol = UnityPal.youngchaGray;

                    if(i.item != null) itemCol = i.item.color;

                    color.r += itemCol.r * liquidVol;
                    color.g += itemCol.g * liquidVol;
                    color.b += itemCol.b * liquidVol;
                }
            }

            if(tLiquid > 0f){
                float invt = 1f / tLiquid;

                Draw.color(color.mul(invt), Mathf.clamp(tLiquid * fraction * 2f));
                Draw.rect(liquidRegions[tIndex], x, y, 8f, 8f);
            }

            for(var i : cc){
                if(i.meltedRatio < 1f && i.volume * fraction > 0.1f){
                    Color itemCol = UnityPal.youngchaGray;

                    if(i.item != null) itemCol = i.item.color;

                    float ddd = (1f - i.meltedRatio) * i.volume * fraction;

                    if(ddd > 0.1f){
                        Draw.color(itemCol);
                        if(ddd > 1f) Draw.rect(solidItemStrips[Mathf.floor(ddd) - 1], x, y);

                        float siz = 8f * (ddd % 1f);
                        long pos = randomPos[Math.max(Mathf.floor(ddd), 5)];

                        Draw.rect(solidItem, SVec2.x(pos) + x, SVec2.y(pos) + y, siz, siz);
                    }
                }
            }

            Draw.color();
        }
    }
}
